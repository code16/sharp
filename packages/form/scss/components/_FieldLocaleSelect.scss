.SharpFieldLocaleSelect {
  margin-bottom: 0.625rem;
  &__btn {
    text-transform: uppercase;
    background: $ui-01;
    border: 1px solid $ui-03;
    color: $text-02;
    border-radius: 5px;
    -webkit-appearance: none;
    outline: 0;
    cursor: pointer;

    &:hover {
      border: 1px solid $brand-01;
    }
  }
  &__btn--empty {
    border: 1px dashed $brand-01;
    background: transparent;
  }
  &__btn--active {
    background: $brand-01;
    color: $inverse-01;
  }
}