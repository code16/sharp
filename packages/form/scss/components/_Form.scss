@import "../mixins/fix-module-fields-margin";

.SharpForm {

  .close {
    font-weight: normal;
  }

  &__fieldset {
    @include reset;
    //margin-bottom: 2rem;

    @extend .SharpModule;
    @include fix-module-fields-margin;

    &>.SharpModule__inner {
      background: transparent;
      border: 0;
      @include layer('base');
      &>.SharpModule__header {
        padding: 0;
        .SharpModule__title {
          @extend .SharpForm__label;
        }
      }
      &>.SharpModule__content {
        background: $ui-03;
        @include layer('raised');
      }
    }
  }

  .SharpGrid__col--empty {
    display: none;
  }

  .SharpGrid__col:not(.SharpGrid__col--empty), &__form-item--row {
    margin-bottom: $form-item-margin-bottom;
  }

  &__form-item {
    flex: 1;
    height: 100%;
    //margin-right: 1rem;

    /*&:last-of-type {
      margin-right: 0;
      margin-bottom: 0;
    }*/

    // For single check align with other field
    &--type-check {
      .SharpForm__label {
        &::after {
          content: "\a0\a0"
        }
      }
      .SharpCheck {
        height: 2.5rem;
      }
    }
  }

  &__label {
    @include reset;
    @include font-size('14');
    font-weight: $input-label-weight;
    display: inline-block;
    vertical-align: baseline;
    margin-bottom: rem(10px) !important;
    text-transform: uppercase;
  }

  &__label--disabled {
    opacity: .5;
  }

  &__form-item--danger {
    input, textarea, select, .SharpDropdown__text {
      //border-bottom: 2px solid $support-01;
    }
    .SharpForm__form-requirement {
      max-height: rem(200px);
      background: rgba($support-01, .1);
      margin: 0;
      padding: .5rem 1rem;
    }
  }


  &__form-requirement {
    @include reset;
    @include font-size('11');
    margin: .75rem 0 0;
    max-height: 0;
    overflow: hidden;
    font-weight: 700;
    line-height: 1.5;

    &:empty {
      display:none;
    }

    &::before {
      content: '*';
      display: inline-block;
      color: $support-01;
    }
  }

  &__help-message {
    @include font-size('11');
    margin: .75rem 0 0;
    line-height: 1.5;
  }



  &__locale-dropdown {
    text-transform: uppercase;
    width: 4rem;
  }

  .SharpTabs {
    .SharpForm__locale-dropdown {
      align-self: center;
      margin-right: 1rem;
      @include media-breakpoint-up(md) {
        position: absolute;
        left: 0;
        top: 0;
        margin-right: 0;
      }
    }
    &--nav-overflow {
      .SharpForm__locale-dropdown {
        order: 1;
        margin-top: .5rem;
        align-self: flex-start;
      }
      .SharpTabs__collapse-btn {
        align-self: flex-start;
      }
    }
  }

  /// reset multiselect new Design
  .multiselect__tags {
    border-radius: 0;
    border: none;
  }

  .multiselect__option:not(:hover):not(.multiselect__option--selected):not(.multiselect__option--highlighted)
  {
      background: white!important;
  }
}