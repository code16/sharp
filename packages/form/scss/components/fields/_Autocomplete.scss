.SharpAutocomplete {
  position: relative;
  font-size: 1rem;

  &--disabled {
    opacity: .5;
    cursor: not-allowed !important;
    outline: none !important;
    .SharpAutocomplete__result-item__close-button {
      pointer-events: none;
      cursor: not-allowed !important;
    }
  }

  &__result-item {
    &__close-button {
      @extend %select-close-button;
    }

    &__close-icon {
      @extend %select-close-icon;
    }
  }

  &__multiselect {
    &--hide-dropdown.multiselect--active {
      .multiselect__content {
        display: none !important;
      }
    }
    .multiselect__tags {
      padding-right: 2rem !important;
    }
    &.multiselect--active {
      .multiselect__input {
        width: 100%!important;
      }
    }
    .multiselect__option > .SharpTemplate {
      overflow:hidden;
      text-overflow: ellipsis;
    }
    .multiselect__single {
      .SharpTemplate {
        overflow:hidden;
        text-overflow: ellipsis;
      }
    }
  }

  &__overlay {
    position: absolute;
    left: 0;
    top: 0;
    right: 2rem;
    bottom: 0;
    width: auto;

    .multiselect__tags {
      padding-right: 0;
      box-shadow: none!important;
      user-select: auto!important;
    }
  }

  &--remote {
    .multiselect__select {
      display: none;
    }

    .multiselect__tags {
      padding-right: 1rem;
    }
  }

  .SharpLoading {
    position: absolute;
    right: 0;
    top: 50%;
    transform: translateY(-50%);
    &__svg {
      stroke: $brand-01
    }
  }
}