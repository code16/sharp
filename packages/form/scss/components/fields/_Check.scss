.SharpCheck {
  height: 100%;
  &__label {
    @include reset;
    @include font-smoothing;
    @include font-size('14');
    display: flex;
    height: 100%;
    align-items: center;
    cursor: pointer;

    .SharpCheck__input {
      @include hidden;
    }

    .SharpCheck__label-text {
      user-select: none;
    }

    .SharpCheck__appearance {
      display: inline-block;
      height: rem(18px);
      min-width: rem(18px);
      max-width: rem(18px);
      margin-right: .5rem;
      background-color: $ui-01;
      border-radius: 2px;
      border: 2px solid $ui-05;
    }

    .SharpCheck__input:checked + .SharpCheck__appearance,
    .SharpCheck__input:checked + .SharpCheck__label .SharpCheck__appearance {
      display: flex;
      align-items: baseline;
      background-color: $brand-01;
      border-color: $brand-01;
    }

    .SharpCheck__input:disabled + .SharpCheck__appearance,
    .SharpCheck__input:disabled ~ .SharpCheck__label-text {
      opacity: .5;
      cursor: not-allowed;
    }

    .SharpCheck__input:focus + .SharpCheck__appearance {
      @include focus-outline('blurred');
    }

    .SharpCheck__checkmark {
      display: none;
      fill: $inverse-01;
      width: rem(15px);
      height: rem(15px);
    }

    .SharpCheck__input:checked + .SharpCheck__appearance .SharpCheck__checkmark {
      display: block;

      @-moz-document url-prefix() {
        & {
          stroke: $brand-01;
        }
      }
    }
  }
}