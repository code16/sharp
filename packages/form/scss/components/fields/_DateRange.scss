@import '~element-ui/packages/theme-chalk/src/date-picker';

.SharpDateRange {
  .el-range__close-icon {
    display: none;
    font-weight: normal;
    font-size: 16px;
    -webkit-font-smoothing: subpixel-antialiased;
    font-family: inherit!important;
    &::before {
      content: 'Ã—';
    }
  }
  &--clearable {
    .el-range__close-icon {
      display: inline-block;
    }
  }
}