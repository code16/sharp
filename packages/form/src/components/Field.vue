<script>
    import { log, logError, isCustomField, resolveCustomField } from 'sharp';
    import Fields from './fields/index';

    export default {
        name:'SharpField',
        inheritAttrs: false,

        provide() {
            return {
                $field: this
            }
        },

        props: {
            fieldKey: String,
            fieldType: String,
            fieldProps: Object,
            fieldLayout: Object,
            value: [String, Number, Boolean, Object, Array, Date],
            locale: String,
            uniqueIdentifier: String,
            fieldConfigIdentifier: String,
            updateData: Function,
            readOnly: Boolean,
        },
        computed: {
            isCustom() {
                return isCustomField(this.fieldType)
            },
            component() {
                if(this.isCustom) {
                    return resolveCustomField(this.fieldType);
                }
                return Fields[this.fieldType];
            }
        },
        render(h) {
            if(!this.component) {
                if(!this.custom) {
                    logError(`Unknown field type '${this.fieldType}'`, this.fieldProps);
                }
                return null;
            }

            let { key, ...fieldProps } = this.fieldProps;


            return h(this.component, {
                props: {
                    fieldKey: this.fieldKey,
                    fieldLayout: this.fieldLayout,
                    value: this.value,
                    locale: this.locale,
                    uniqueIdentifier: this.uniqueIdentifier,
                    fieldConfigIdentifier: this.fieldConfigIdentifier,
                    ...fieldProps
                },
                attrs: {
                    dynamicAttributes: fieldProps.dynamicAttributes,
                },
                on: {
                    input: (val, options={}) => {
                        if(this.fieldProps.readOnly && !options.force)
                            log(`SharpField '${this.fieldKey}', can't update because is readOnly`);
                        else
                            this.updateData(this.fieldKey, val, { forced:options.force });
                    },
                    blur: () => {
                        this.fieldProps.focused = false;
                    }
                }
            });
        }
    }
</script>