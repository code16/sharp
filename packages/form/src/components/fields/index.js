import Autocomplete from './Autocomplete.vue';
import Textarea from './Textarea.vue';
import Text from './Text.vue';
import Markdown from './markdown/Markdown.vue';
import NumberInput from './Number.vue';
import Upload from './upload/Upload.vue';
import TagInput from './Tags.vue';
import DateInput from './date/Date.vue';
import Check from './Check.vue';
import List from './list/List.vue';
import Select from './Select.vue';
import Html from './Html.vue';
import Geolocation from './geolocation/Geolocation';
import Trix from './wysiwyg/TrixEditor.vue';
import DateRange from './date-range/DateRange';

export default {
    'autocomplete' : Autocomplete,
    'text'         : Text,
    'textarea'     : Textarea,
    'markdown'     : Markdown,
    'number'       : NumberInput,
    'upload'       : Upload,
    'tags'         : TagInput,
    'date'         : DateInput,
    'check'        : Check,
    'list'         : List,
    'select'       : Select,
    'html'         : Html,
    'geolocation'  : Geolocation,
    'wysiwyg'      : Trix,
    'daterange'    : DateRange,
};

export {
    Autocomplete,
    Text,
    Textarea,
    Markdown,
    NumberInput,
    Upload,
    TagInput,
    DateInput,
    Check,
    List,
    Select,
    Html,
    Geolocation,
    Trix,
    DateRange,
}