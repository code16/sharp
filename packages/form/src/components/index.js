
export * from './fields';
export { default as Form } from './Form';