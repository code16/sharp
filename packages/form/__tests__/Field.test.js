import Vue from 'vue';
import Field from '../src/components/Field.vue';
import * as fields from '../src/components/fields';

import { logError } from 'sharp';

jest.mock('sharp');

describe('sharp-field', () => {
    Vue.component('sharp-field', Field);

    beforeEach(() => {
        document.body.innerHTML = `
            <div id="app">
                <sharp-field field-key="title" 
                :field-type="fieldType" 
                :field-props="fieldProps" 
                :field-layout="fieldLayout" 
                :value="value"
                unique-identifier="list.0.title"
                field-config-identifier="list.title"
                :update-data="updateData">
                </sharp-field>      
            </div>
        `;
    });

    test("can't mount and display error if unknown field", async () => {
        await createVm({
            propsData: {
                fieldType: 'unknown'
            }
        });

        expect(logError).toHaveBeenCalled();
        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount field', async () => {
        fields.default['test-field'] = {
            template: '<div>Template field</div>'
        };

        await createVm({
            propsData: {
                fieldType: 'test-field',
                fieldProps: {}
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('expose proper props', async () => {
        let $field = await createVm({
            propsData: {
                fieldType: 'text',
                fieldProps: {
                    placeholder: 'Titre'
                },
                fieldLayout: {
                    tabs: []
                }
            },
            data:()=>({ value: 'Hello' })
        });

        let { $children:[renderedField] } = $field;

        expect(renderedField.$vnode.data.props).toEqual({
            fieldKey: 'title',
            fieldLayout: {
                tabs: []
            },
            value: 'Hello',
            placeholder: 'Titre',
            uniqueIdentifier: 'list.0.title',
            fieldConfigIdentifier: 'list.title'
        });
    });

    test('call update if input emitted', async () => {
        let updateData = jest.fn();
        let $field = await createVm({
            propsData: {
                fieldType: 'text',
                fieldProps: {}
            },
            methods: {
                updateData
            }
        });

        let { $children:[renderedField] } = $field;

        renderedField.$emit('input', 'Coucou');

        expect(updateData).toHaveBeenCalledTimes(1);
        expect(updateData).toHaveBeenCalledWith('title','Coucou', { forced:undefined });

        updateData.mockClear();
        renderedField.$emit('input', 'Bonjour', { force: true });

        expect(updateData).toHaveBeenCalledTimes(1);
        expect(updateData).toHaveBeenCalledWith('title', 'Bonjour', { forced: true });
    });

    test("don't call update when readOnly if input emitted", async () => {
        let updateData = jest.fn();
        let $field = await createVm({
            propsData: {
                fieldType: 'text',
                fieldProps: {
                    readOnly: true
                }
            },
            methods: {
                updateData
            }
        });

        let { $children:[renderedField] } = $field;

        renderedField.$emit('input', 'Coucou');

        expect(updateData).not.toHaveBeenCalled();

        renderedField.$emit('input', 'Coucou', { force: true });

        expect(updateData).toHaveBeenCalled();

    });

    test('provide "$field" injection', async () => {
        fields.default['test-field'] = {
            template: '<div>Template field</div>',
            inject:['$field']
        };
        let $field = await createVm({
            propsData: {
                fieldType: 'test-field',
                fieldProps: {}
            },
        });

        let { $children:[renderedField] } = $field;

        expect(renderedField.$field).toBe($field);
    });
});

async function createVm(customOptions={}) {

    const vm = new Vue({
        el: '#app',
        mixins: [customOptions],

        props:['fieldType', 'fieldProps', 'fieldLayout'],

        'extends': {
            methods: {
                updateData:()=>{}
            },
            data:()=>({
                value: null
            })
        }
    });

    await Vue.nextTick();

    return vm.$children[0];
}