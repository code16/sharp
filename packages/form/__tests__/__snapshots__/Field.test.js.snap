// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sharp-field can mount field 1`] = `
<div id="app">
	<div>Template field</div>
</div>
`;

exports[`sharp-field can't mount and display error if unknown field 1`] = `
<div id="app">
	<!---->
</div>
`;
